python/src/iceberg/types.py

```diff
@@ -15,61 +15,84 @@
 # specific language governing permissions and limitations
 # under the License.
 """Data types used in describing Iceberg schemas
-
 This module implements the data types described in the Iceberg specification for Iceberg schemas. To
 describe an Iceberg table schema, these classes can be used in the construction of a StructType instance.
-
 Example:
     >>> StructType(
         [
             NestedField(True, 1, "required_field", StringType()),
             NestedField(False, 2, "optional_field", IntegerType()),
         ]
     )
-
 Notes:
   - https://iceberg.apache.org/#spec/#primitive-types
 """
 
-from typing import Optional
+from typing import Any, Dict, List, Optional, Tuple
 
 
-class Type:
-    def __init__(self, type_string: str, repr_string: str, is_primitive=False):
-        self._type_string = type_string
-        self._repr_string = repr_string
-        self._is_primitive = is_primitive
+class IcebergType:
+
+    _implemented: Dict[Tuple[str, Tuple[Any]], "IcebergType"] = {}
+
+    def __new__(cls, *args, **kwargs):
+        init_func = cls.__dict__.get("__init__")
+        names = init_func.__code__.co_varnames[1:] if init_func else []
+        defaults = init_func and init_func.__defaults__ or []
+        args = {**dict(zip(names[::-1], defaults[::-1])), **dict(zip(names, args))}
+        args = {**args, **kwargs}
+        args = tuple(
+            (k, tuple(args[k]) if isinstance(args[k], list) else args[k]) for k in names
+        )
+        key = (cls.__name__, args)
+        if key in cls._implemented:
+            return cls._implemented[key]
+        else:
+            ret = object.__new__(cls)
+            ret._args = args
+            cls._implemented[key] = ret
+            return ret
 
     def __repr__(self):
-        return self._repr_string
+        base = f"{type(self).__name__}("
+        if not self._args:
+            return base + ")"
+        for k, v in self._args:
+            base += k + "=" + repr(v) + ", "
+        return base[:-2] + ")"
```

Comment:
You can likely use `join` and a comprehension so you don't have to chop off the last two characters of the string.

Untested, but maybe something like
```python
', '.join([f"{k=repr(v)" for k, v in self._args])
```

Reply:
thanks changed it. definitely makes that more concise